File: lib/widgets/docked_player.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:cached_network_image/cached_network_image.dart';

import '../models/livetv_channel.dart';
import '../mpv/mpv.dart';
import '../services/last_watched_service.dart';
import '../services/settings_service.dart';
import '../utils/app_logger.dart';

/// Docked/mini player widget for the home screen
/// Displays the last watched channel in a compact player at the top of the screen
/// Similar to DirecTV's home screen player that auto-plays last watched content
class DockedPlayer extends StatefulWidget {
  /// Callback when user taps to expand to fullscreen
  final VoidCallback? onExpandToFullscreen;

  /// Callback when user changes the channel
  final void Function(LiveTVChannel)? onChannelChange;

  /// List of available channels for navigation
  final List<LiveTVChannel> channels;

  /// Whether the player should be visible
  final bool isVisible;

  /// Whether to auto-play on mount
  final bool autoPlay;

  /// Height of the docked player
  final double height;

  const DockedPlayer({
    super.key,
    this.onExpandToFullscreen,
    this.onChannelChange,
    this.channels = const [],
    this.isVisible = true,
    this.autoPlay = true,
    this.height = 200,
  });

  @override
  State<DockedPlayer> createState() => DockedPlayerState();
}

class DockedPlayerState extends State<DockedPlayer> with WidgetsBindingObserver {
  Player? _player;
  bool _isPlayerInitialized = false;
  LiveTVChannel? _currentChannel;
  final ValueNotifier<bool> _isBuffering = ValueNotifier<bool>(true);
  StreamSubscription<bool>? _bufferingSubscription;
  String? _errorMessage;
  bool _isMuted = false;
  bool _showControls = false;
  Timer? _controlsTimer;
  final FocusNode _focusNode = FocusNode(debugLabel: 'DockedPlayer');

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);
    if (widget.autoPlay) {
      _loadLastWatchedChannel();
    }
  }

  @override
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    _bufferingSubscription?.cancel();
    _controlsTimer?.cancel();
    _player?.dispose();
    _isBuffering.dispose();
    _focusNode.dispose();
    super.dispose();
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    if (state == AppLifecycleState.paused) {
      _player?.pause();
    } else if (state == AppLifecycleState.resumed && _currentChannel != null) {
      _player?.play();
    }
  }

  Future<void> _loadLastWatchedChannel() async {
    try {
      final lastWatchedService = await LastWatchedService.getInstance();
      final lastChannel = lastWatchedService.getLastWatchedChannel();

      if (lastChannel != null && mounted) {
        setState(() {
          _currentChannel = lastChannel;
        });
        await _initializePlayer();
        await _playChannel(lastChannel);
      }
    } catch (e) {
      appLogger.e('Failed to load last watched channel', error: e);
    }
  }

  Future<void> _initializePlayer() async {
    if (_isPlayerInitialized) return;

    try {
      final settingsService = await SettingsService.getInstance();
      final bufferSizeMB = settingsService.getBufferSize();
      final bufferSizeBytes = bufferSizeMB * 1024 * 1024;
      final enableHardwareDecoding = settingsService.getEnableHardwareDecoding();

      _player = Player();

      // Configure player for live streaming (lower latency for docked player)
      await _player!.setProperty('demuxer-max-bytes', bufferSizeBytes.toString());
      await _player!.setProperty('hwdec', enableHardwareDecoding ? 'auto-safe' : 'no');
      await _player!.setProperty('cache', 'yes');
      await _player!.setProperty('cache-secs', '10');
      await _player!.setProperty('demuxer-readahead-secs', '10');
      await _player!.setProperty(
        'stream-lavf-o',
        'reconnect=1,reconnect_streamed=1,reconnect_delay_max=2',
      );

      // Start muted for docked player (user can unmute)
      await _player!.setVolume(0);

      _bufferingSubscription = _player!.streams.buffering.listen((buffering) {
        _isBuffering.value = buffering;
      });

      setState(() {
        _isPlayerInitialized = true;
        _isMuted = true;
      });
    } catch (e) {
      appLogger.e('Failed to initialize docked player', error: e);
      setState(() {
        _errorMessage = 'Failed to initialize player';
      });
    }
  }

  Future<void> _playChannel(LiveTVChannel channel) async {
    if (_player == null) return;

    try {
      setState(() {
        _currentChannel = channel;
        _errorMessage = null;
      });

      await _player!.open(Media(channel.streamUrl));

      // Update last watched service
      final lastWatchedService = await LastWatchedService.getInstance();
      await lastWatchedService.setLastWatchedChannel(channel);

      widget.onChannelChange?.call(channel);
    } catch (e) {
      appLogger.e('Failed to play channel: ${channel.name}', error: e);
      setState(() {
        _errorMessage = 'Failed to play ${channel.name}';
      });
    }
  }

  /// Play a specific channel
  Future<void> playChannel(LiveTVChannel channel) async {
    if (!_isPlayerInitialized) {
      await _initializePlayer();
    }
    await _playChannel(channel);
  }

  /// Stop playback
  void stop() {
    _player?.stop();
    setState(() {
      _currentChannel = null;
    });
  }

  /// Toggle mute
  void toggleMute() async {
    if (_player == null) return;

    final settingsService = await SettingsService.getInstance();
    final savedVolume = settingsService.getVolume();

    if (_isMuted) {
      await _player!.setVolume(savedVolume);
    } else {
      await _player!.setVolume(0);
    }

    setState(() {
      _isMuted = !_isMuted;
    });
  }

  void _showControlsTemporarily() {
    setState(() {
      _showControls = true;
    });
    _controlsTimer?.cancel();
    _controlsTimer = Timer(const Duration(seconds: 3), () {
      if (mounted) {
        setState(() {
          _showControls = false;
        });
      }
    });
  }

  KeyEventResult _handleKeyEvent(FocusNode node, KeyEvent event) {
    if (event is! KeyDownEvent) return KeyEventResult.ignored;

    // Enter to expand to fullscreen
    if (event.logicalKey == LogicalKeyboardKey.enter ||
        event.logicalKey == LogicalKeyboardKey.select) {
      widget.onExpandToFullscreen?.call();
      return KeyEventResult.handled;
    }

    // M or Space to toggle mute
    if (event.logicalKey == LogicalKeyboardKey.keyM ||
        event.logicalKey == LogicalKeyboardKey.space) {
      toggleMute();
      _showControlsTemporarily();
      return KeyEventResult.handled;
    }

    // Show controls on any other key
    _showControlsTemporarily();

    return KeyEventResult.ignored;
  }

  @override
  Widget build(BuildContext context) {
    if (!widget.isVisible || _currentChannel == null) {
      return const SizedBox.shrink();
    }

    return Focus(
      focusNode: _focusNode,
      onKeyEvent: _handleKeyEvent,
      child: GestureDetector(
        onTap: () {
          _showControlsTemporarily();
          _focusNode.requestFocus();
        },
        onDoubleTap: widget.onExpandToFullscreen,
        child: MouseRegion(
          onEnter: (_) => _showControlsTemporarily(),
          child: AnimatedContainer(
            duration: const Duration(milliseconds: 300),
            height: widget.height,
            decoration: BoxDecoration(
              color: Colors.black,
              borderRadius: BorderRadius.circular(12),
              boxShadow: [
                BoxShadow(
                  color: Colors.black.withValues(alpha: 0.5),
                  blurRadius: 20,
                  offset: const Offset(0, 4),
                ),
              ],
            ),
            clipBehavior: Clip.antiAlias,
            child: Stack(
              fit: StackFit.expand,
              children: [
                // Video player
                if (_isPlayerInitialized && _player != null)
                  Video(player: _player!)
                else
                  Container(color: Colors.black),

                // Buffering indicator
                ValueListenableBuilder<bool>(
                  valueListenable: _isBuffering,
                  builder: (context, isBuffering, child) {
                    if (!isBuffering) return const SizedBox.shrink();
                    return Container(
                      color: Colors.black54,
                      child: const Center(
                        child: CircularProgressIndicator(
                          valueColor: AlwaysStoppedAnimation<Color>(Colors.white),
                        ),
                      ),
                    );
                  },
                ),

                // Error message
                if (_errorMessage != null)
                  Container(
                    color: Colors.black87,
                    child: Center(
                      child: Column(
                        mainAxisSize: MainAxisSize.min,
                        children: [
                          const Icon(Icons.error_outline, color: Colors.red, size: 32),
                          const SizedBox(height: 8),
                          Text(
                            _errorMessage!,
                            style: const TextStyle(color: Colors.white),
                            textAlign: TextAlign.center,
                          ),
                        ],
                      ),
                    ),
                  ),

                // Gradient overlay for controls
                AnimatedOpacity(
                  opacity: _showControls ? 1.0 : 0.0,
                  duration: const Duration(milliseconds: 200),
                  child: Container(
                    decoration: BoxDecoration(
                      gradient: LinearGradient(
                        begin: Alignment.topCenter,
                        end: Alignment.bottomCenter,
                        colors: [
                          Colors.black.withValues(alpha: 0.7),
                          Colors.transparent,
                          Colors.transparent,
                          Colors.black.withValues(alpha: 0.8),
                        ],
                        stops: const [0.0, 0.3, 0.7, 1.0],
                      ),
                    ),
                  ),
                ),

                // "LIVE" badge
                Positioned(
                  top: 12,
                  left: 12,
                  child: Container(
                    padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 4),
                    decoration: BoxDecoration(
                      color: Colors.red,
                      borderRadius: BorderRadius.circular(4),
                    ),
                    child: const Text(
                      'LIVE',
                      style: TextStyle(
                        color: Colors.white,
                        fontSize: 10,
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                  ),
                ),

                // Mute indicator
                if (_isMuted)
                  Positioned(
                    top: 12,
                    right: 12,
                    child: Container(
                      padding: const EdgeInsets.all(6),
                      decoration: BoxDecoration(
                        color: Colors.black54,
                        borderRadius: BorderRadius.circular(20),
                      ),
                      child: const Icon(
                        Icons.volume_off,
                        color: Colors.white,
                        size: 18,
                      ),
                    ),
                  ),

                // Bottom info bar
                Positioned(
                  left: 0,
                  right: 0,
                  bottom: 0,
                  child: AnimatedOpacity(
                    opacity: _showControls ? 1.0 : 0.7,
                    duration: const Duration(milliseconds: 200),
                    child: Container(
                      padding: const EdgeInsets.all(12),
                      child: Row(
                        children: [
                          // Channel logo
                          if (_currentChannel?.logo != null)
                            Container(
                              width: 40,
                              height: 40,
                              margin: const EdgeInsets.only(right: 12),
                              decoration: BoxDecoration(
                                color: Colors.white,
                                borderRadius: BorderRadius.circular(8),
                              ),
                              clipBehavior: Clip.antiAlias,
                              child: CachedNetworkImage(
                                imageUrl: _currentChannel!.logo!,
                                fit: BoxFit.contain,
                                errorWidget: (context, url, error) => Icon(
                                  Icons.tv,
                                  color: Colors.grey[600],
                                ),
                              ),
                            ),
                          // Channel info
                          Expanded(
                            child: Column(
                              crossAxisAlignment: CrossAxisAlignment.start,
                              mainAxisSize: MainAxisSize.min,
                              children: [
                                Text(
                                  _currentChannel?.name ?? '',
                                  style: const TextStyle(
                                    color: Colors.white,
                                    fontWeight: FontWeight.bold,
                                    fontSize: 14,
                                  ),
                                  maxLines: 1,
                                  overflow: TextOverflow.ellipsis,
                                ),
                                if (_currentChannel?.nowPlaying != null) ...[
                                  const SizedBox(height: 2),
                                  Text(
                                    _currentChannel!.nowPlaying!.title,
                                    style: TextStyle(
                                      color: Colors.white.withValues(alpha: 0.8),
                                      fontSize: 12,
                                    ),
                                    maxLines: 1,
                                    overflow: TextOverflow.ellipsis,
                                  ),
                                ],
                              ],
                            ),
                          ),
                          // Expand button
                          if (_showControls)
                            IconButton(
                              icon: const Icon(
                                Icons.fullscreen,
                                color: Colors.white,
                              ),
                              onPressed: widget.onExpandToFullscreen,
                              tooltip: 'Watch fullscreen',
                            ),
                          // Mute button
                          if (_showControls)
                            IconButton(
                              icon: Icon(
                                _isMuted ? Icons.volume_off : Icons.volume_up,
                                color: Colors.white,
                              ),
                              onPressed: toggleMute,
                              tooltip: _isMuted ? 'Unmute' : 'Mute',
                            ),
                        ],
                      ),
                    ),
                  ),
                ),

                // Focus indicator
                if (_focusNode.hasFocus)
                  Positioned.fill(
                    child: Container(
                      decoration: BoxDecoration(
                        borderRadius: BorderRadius.circular(12),
                        border: Border.all(
                          color: Theme.of(context).colorScheme.primary,
                          width: 3,
                        ),
                      ),
                    ),
                  ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

File: android-native/app/src/main/java/com/openflix/presentation/components/DockedPlayer.kt
package com.openflix.presentation.components

import android.view.SurfaceView
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.key.*
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.viewinterop.AndroidView
import androidx.tv.material3.*
import coil.compose.AsyncImage
import com.openflix.domain.model.Channel
import com.openflix.player.LiveTVPlayer
import com.openflix.presentation.theme.OpenFlixColors
import kotlinx.coroutines.delay
import timber.log.Timber

/**
 * DirecTV-style docked player for the home screen.
 * Auto-plays the last watched channel in a compact preview.
 * Similar to DirecTV's home screen that shows live TV in corner.
 */
@Composable
fun DockedPlayer(
    channel: Channel?,
    liveTVPlayer: LiveTVPlayer,
    isVisible: Boolean = true,
    onExpandToFullscreen: () -> Unit,
    onChannelChange: (Channel) -> Unit = {},
    modifier: Modifier = Modifier
) {
    if (!isVisible || channel == null || channel.streamUrl == null) {
        return
    }

    var isFocused by remember { mutableStateOf(false) }
    var showControls by remember { mutableStateOf(false) }
    val focusRequester = remember { FocusRequester() }

    // Auto-hide controls after 3 seconds
    LaunchedEffect(showControls) {
        if (showControls) {
            delay(3000)
            showControls = false
        }
    }

    // Player states
    val isPlaying by liveTVPlayer.isPlaying.collectAsState()
    val isBuffering by liveTVPlayer.isBuffering.collectAsState()
    val isMuted by liveTVPlayer.isMuted.collectAsState()
    val error by liveTVPlayer.error.collectAsState()

    // Initialize and play on mount
    LaunchedEffect(channel.streamUrl) {
        Timber.d("DockedPlayer: Starting playback of ${channel.name}")
        liveTVPlayer.initialize()
        liveTVPlayer.setMuted(true) // Start muted
        liveTVPlayer.play(channel.streamUrl!!)
    }

    // Focus border animation
    val borderAlpha by animateFloatAsState(
        targetValue = if (isFocused) 1f else 0f,
        label = "borderAlpha"
    )

    Surface(
        onClick = {
            showControls = !showControls
        },
        modifier = modifier
            .fillMaxWidth()
            .height(200.dp)
            .focusRequester(focusRequester)
            .onFocusChanged {
                isFocused = it.isFocused
                if (it.isFocused) showControls = true
            }
            .onPreviewKeyEvent { event ->
                if (event.type == KeyEventType.KeyDown) {
                    when (event.key) {
                        Key.Enter, Key.DirectionCenter -> {
                            onExpandToFullscreen()
                            true
                        }
                        Key.M, Key.VolumeMute -> {
                            liveTVPlayer.toggleMute()
                            showControls = true
                            true
                        }
                        Key.Spacebar -> {
                            liveTVPlayer.togglePlayPause()
                            showControls = true
                            true
                        }
                        else -> false
                    }
                } else false
            },
        shape = ClickableSurfaceDefaults.shape(RoundedCornerShape(16.dp)),
        colors = ClickableSurfaceDefaults.colors(
            containerColor = Color.Black,
            focusedContainerColor = Color.Black
        ),
        border = ClickableSurfaceDefaults.border(
            focusedBorder = Border(
                border = BorderStroke(3.dp, OpenFlixColors.Primary),
                shape = RoundedCornerShape(16.dp)
            )
        ),
        scale = ClickableSurfaceDefaults.scale(focusedScale = 1.02f)
    ) {
        Box(modifier = Modifier.fillMaxSize()) {
            // Video Surface
            AndroidView(
                factory = { context ->
                    SurfaceView(context).also { surface ->
                        liveTVPlayer.setSurfaceView(surface)
                    }
                },
                modifier = Modifier.fillMaxSize(),
                onRelease = {
                    // Don't release the player, just detach surface
                    liveTVPlayer.setSurfaceView(null)
                }
            )

            // Buffering indicator
            if (isBuffering) {
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .background(Color.Black.copy(alpha = 0.5f)),
                    contentAlignment = Alignment.Center
                ) {
                    Box(
                        modifier = Modifier
                            .size(40.dp)
                            .clip(CircleShape)
                            .background(OpenFlixColors.Primary.copy(alpha = 0.3f)),
                        contentAlignment = Alignment.Center
                    ) {
                        Box(
                            modifier = Modifier
                                .size(24.dp)
                                .clip(CircleShape)
                                .background(OpenFlixColors.Primary)
                        )
                    }
                }
            }

            // Error overlay
            if (error != null) {
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .background(Color.Black.copy(alpha = 0.8f)),
                    contentAlignment = Alignment.Center
                ) {
                    Column(
                        horizontalAlignment = Alignment.CenterHorizontally,
                        verticalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        Icon(
                            imageVector = Icons.Filled.Error,
                            contentDescription = null,
                            tint = OpenFlixColors.Error,
                            modifier = Modifier.size(32.dp)
                        )
                        Text(
                            text = "Unable to play",
                            style = MaterialTheme.typography.bodyMedium,
                            color = OpenFlixColors.TextPrimary
                        )
                    }
                }
            }

            // Gradient overlays
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .background(
                        Brush.verticalGradient(
                            colors = listOf(
                                Color.Black.copy(alpha = 0.6f),
                                Color.Transparent,
                                Color.Transparent,
                                Color.Black.copy(alpha = 0.8f)
                            ),
                            startY = 0f,
                            endY = Float.POSITIVE_INFINITY
                        )
                    )
            )

            // LIVE Badge
            Box(
                modifier = Modifier
                    .padding(12.dp)
                    .align(Alignment.TopStart)
                    .background(Color.Red, RoundedCornerShape(4.dp))
                    .padding(horizontal = 8.dp, vertical = 4.dp)
            ) {
                Text(
                    text = "LIVE",
                    style = MaterialTheme.typography.labelSmall,
                    fontWeight = FontWeight.Bold,
                    color = Color.White,
                    letterSpacing = 1.sp
                )
            }

            // Mute indicator
            if (isMuted) {
                Box(
                    modifier = Modifier
                        .padding(12.dp)
                        .align(Alignment.TopEnd)
                        .background(Color.Black.copy(alpha = 0.6f), CircleShape)
                        .padding(6.dp)
                ) {
                    Icon(
                        imageVector = Icons.Filled.VolumeOff,
                        contentDescription = "Muted",
                        tint = Color.White,
                        modifier = Modifier.size(18.dp)
                    )
                }
            }

            // Bottom info bar
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .align(Alignment.BottomCenter)
                    .padding(12.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                // Channel logo
                channel.logoUrl?.let { logoUrl ->
                    Box(
                        modifier = Modifier
                            .size(44.dp)
                            .clip(RoundedCornerShape(8.dp))
                            .background(Color.White)
                            .padding(4.dp)
                    ) {
                        AsyncImage(
                            model = logoUrl,
                            contentDescription = channel.name,
                            modifier = Modifier.fillMaxSize(),
                            contentScale = ContentScale.Fit
                        )
                    }
                    Spacer(modifier = Modifier.width(12.dp))
                }

                // Channel info
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        text = channel.displayName,
                        style = MaterialTheme.typography.titleMedium,
                        fontWeight = FontWeight.Bold,
                        color = Color.White,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis
                    )
                    channel.nowPlaying?.let { program ->
                        Text(
                            text = program.title,
                            style = MaterialTheme.typography.bodySmall,
                            color = Color.White.copy(alpha = 0.8f),
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis
                        )
                    }
                }

                // Control buttons (visible on focus)
                AnimatedVisibility(
                    visible = showControls || isFocused,
                    enter = fadeIn(),
                    exit = fadeOut()
                ) {
                    Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                        // Mute button
                        IconButton(
                            onClick = { liveTVPlayer.toggleMute() }
                        ) {
                            Icon(
                                imageVector = if (isMuted) Icons.Filled.VolumeOff else Icons.Filled.VolumeUp,
                                contentDescription = if (isMuted) "Unmute" else "Mute",
                                tint = Color.White
                            )
                        }
                        // Fullscreen button
                        IconButton(
                            onClick = onExpandToFullscreen
                        ) {
                            Icon(
                                imageVector = Icons.Filled.Fullscreen,
                                contentDescription = "Fullscreen",
                                tint = Color.White
                            )
                        }
                    }
                }
            }

            // Now playing progress bar
            channel.nowPlaying?.let { program ->
                val progress = program.progress
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(3.dp)
                        .align(Alignment.BottomCenter)
                        .background(Color.White.copy(alpha = 0.3f))
                ) {
                    Box(
                        modifier = Modifier
                            .fillMaxHeight()
                            .fillMaxWidth(progress)
                            .background(OpenFlixColors.Primary)
                    )
                }
            }
        }
    }
}
